### PR TITLE
Fixing #1288

### PR DESCRIPTION
**Pull request by WhiteBlackGoose**

### CODE REVIEW

**Comment by WhiteBlackGoose** [comment]
Oh, alright, forgot that I can't comment on snippets that I haven't changed.

I found the place where it cannot parse `//`. [This](https://github.com/WhiteBlackGoose/interactive/blob/59f98f0d840566bb25a228fb891d454429f12ede/src/Microsoft.DotNet.Interactive/Parsing/SubmissionParser.cs#L267) thing creates a parser, which I cannot debug [here](https://github.com/WhiteBlackGoose/interactive/blob/59f98f0d840566bb25a228fb891d454429f12ede/src/Microsoft.DotNet.Interactive/Parsing/DirectiveNode.cs#L35).

After this thing parses the given submit code, it returns three errors and three unmatched tokens (`//`, `ducks`, `best!`). Since I don't know what's inside `commandLineBuilder` and how to debug, could you please give some tips on it? @jonsequitur, @colombod.

**Comment by jonsequitur** [comment]
That parser is implemented by the System.CommandLine library. The repo is here: https://github.com/dotnet/command-line-api.

**Comment by WhiteBlackGoose** [comment]
So what do we do? It seems to be an expected behaviour for that library, but doesn't seem so for the `r` magic command. I can think of a workaround, but that's probably not we want.

**Comment by jonsequitur** [comment]
We could potentially trim off the comment before passing it to the parser here: https://github.com/dotnet/interactive/blob/9d899fb9fc1643210b85b64cce727b1b8319c416/src/Microsoft.DotNet.Interactive/Parsing/DirectiveNode.cs#L35

**Comment by WhiteBlackGoose** [comment]
> The thing to keep in mind is that // is perfectly valid within the magic command itself, e.g. #!value --from-url https://example.com.

That's exactly why I'd call trimming as a workaround. Alright, thanks, I'll think

**Comment by WhiteBlackGoose** [comment]
I noticed that you rely on an external `Parser`. I think it might be a good idea to have a wrapper over it, since this `Parser` is only used in the context of directive parsers, this wrapper is `DirectiveParser`. It has a private field of type `Parser`. Now, we can customize its behaviour without ugly workarounds. It also imho looks better (since fields like `directiveParser` are now exactly of the `DirectiveParser` type).

@jonsequitur I would like to hear your thoughts.

File: src/Microsoft.DotNet.Interactive/Parsing/DirectiveParser.cs [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        }
+
+        public ParseResult Parse(string input)
+        {
+            if (FindComment(input) is { } realPos)
+                input = input.Substring(0, realPos);
+            return _commandLineParser.Parse(input);
+        }
+
+        private static int? FindComment(string input)
```
Comment (WhiteBlackGoose) [inline src/Microsoft.DotNet.Interactive/Parsing/DirectiveParser.cs:40]: Oh, and unlike the rest if the code, this part... well, it might be implemented better than I did, so any recommendations are welcomed.

File: src/Microsoft.DotNet.Interactive.Jupyter.Tests/MagicCommandTests.lsmagic.cs [hunk 2]
```diff
@@ -115,5 +118,23 @@ public async Task lsmagic_does_not_list_hidden_commands()
                   .Should()
                   .NotContain("#!hidden");
         }
+
+        [Fact]
+        public async Task r_executes_even_with_comment()
```
Comment (jonsequitur) [inline src/Microsoft.DotNet.Interactive.Jupyter.Tests/MagicCommandTests.lsmagic.cs:123]: Rather than this more integrated test, it would make sense to test `DirectiveParser` directly.

Comment (WhiteBlackGoose) [reply src/Microsoft.DotNet.Interactive.Jupyter.Tests/MagicCommandTests.lsmagic.cs:123]: This makes sense, thanks.

File: src/Microsoft.DotNet.Interactive/Parsing/DirectiveParser.cs [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+            var secondQuote = input.IndexOf('"', firstQuote + 1);
+            if (secondQuote is -1)
+                return null;
+
+            var doubleSlash = input.IndexOf("//", secondQuote + 1);
+            if (doubleSlash is -1)
+                return null;
+
+            return doubleSlash;
+        }
```
Comment (jonsequitur) [inline src/Microsoft.DotNet.Interactive/Parsing/DirectiveParser.cs:55]: It would be useful to start with a few test cases, possibly using `[Theory]`. Quote-counting might not be valuable, since it's a) not required for `#r`, e.g. `#r nuget:package` is valid, and b) it's valid to surround any token with quotes, and it's commonly used to indicate a single token when there's a space in it, e.g. `#!blah "this is parsed as one token"`.

As for test cases, none of the following should be identified as containing a comment:

```console
#!value --from-file //volume/path/to/file
#!value --from-url https://example.com
```

Comment (WhiteBlackGoose) [reply src/Microsoft.DotNet.Interactive/Parsing/DirectiveParser.cs:55]: I thought they aren't directives. If they are, the thing becomes way less trivial. Here's what I see as possible solutions:
- Make a wrapper around `ParseResult` too. The reason is that after we got a `ParseResult` from the command line parser, we need to see if the unmatched tokens are in fact comments. Since `ParseResult` is immutable (which is, in fact, good), we need to extract the necessary data from it, draw our own conclusions regarding comments, and then return our `DirectiveParseResult`.
- Change the behaviour. Make `#r nuget:DuckPackage //quack` raise an error as with invalid syntax (now it swallows errors with no feedback).

What do you think?

Comment (jonsequitur) [reply src/Microsoft.DotNet.Interactive/Parsing/DirectiveParser.cs:55]: Magic commands and directives are the same thing at the implementation level. If there's any semantic distinction it's that a "directive" refers to something that the compiler is or someday will be aware of, i.e. `#i` and `#r`. (Support for these is currently not at parity across C# and F#.)

Since a comment like this is valid for a directive in the underlying language, I think it's useful to fix this. A more targeted fix, just for `#i` and `#r`, might be easier. Distinguishing a comment from valid inputs above will still be challenging though.

Comment (WhiteBlackGoose) [reply src/Microsoft.DotNet.Interactive/Parsing/DirectiveParser.cs:55]: Given that the command line parser can distinguish, I can reinterpret its parse result and see, if the first unmatched token starts with `//`. Okay, let me see if I actually can. I will prepare a number of sets, and you will say if you think that's the behaviour you'd expect.

File: src/Microsoft.DotNet.Interactive.Tests/Parsing/DirectiveParserTests.cs [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+using System.Threading.Tasks;
+using Xunit;
+using System;
+using System.CommandLine;
+
+namespace Microsoft.DotNet.Interactive.Tests.Parsing
+{
+    public class DirectiveParserTests
+    {
+        private static Command GetRDirective()
```
Comment (WhiteBlackGoose) [inline src/Microsoft.DotNet.Interactive.Tests/Parsing/DirectiveParserTests.cs:17]: `r` is private, so I can make it internal and make the assembly visible to the test. Another option is to test `SubmissionParser`, but it's a bit more complicated and high-level than `DirectiveParser`

File: src/Microsoft.DotNet.Interactive/Parsing/DirectiveParser.cs [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+            commandLineBuilder.EnableDirectives = false;
+            _commandLineParser = commandLineBuilder.Build();
+        }
+
+        public DirectiveParseResult Parse(string input)
+        {
+            var parseResult = _commandLineParser.Parse(input);
+            IReadOnlyList<string> unmatchedTokens;
+            IReadOnlyCollection<ParseError> errors;
+            if (parseResult.UnmatchedTokens.Count == parseResult.Errors.Count && parseResult.UnmatchedTokens.Count > 0 && parseResult.UnmatchedTokens[0].StartsWith("//"))
```
Comment (WhiteBlackGoose) [inline src/Microsoft.DotNet.Interactive/Parsing/DirectiveParser.cs:59]: @jonsequitur is it a bad solution, or *very* bad? By this trick when I get as many errors as unmatched tokens, all those errors look like `Unrecognized token bla-bla`, so we can check if the first token starts with `//`.

Comment (jonsequitur) [reply src/Microsoft.DotNet.Interactive/Parsing/DirectiveParser.cs:59]: Interesting solution. It's not too, too bad, given the alternative would probably be to introduce a whole other parser.

This could be tightened up further by making sure that the unmatched tokens all come after the matched tokens. It's possible, I think, for there to be unmatched tokens in between matched tokens.

File: src/Microsoft.DotNet.Interactive.Jupyter.Tests/MagicCommandTests.lsmagic.cs (hunk 2 above)
Comment (jonsequitur) [reply src/Microsoft.DotNet.Interactive.Jupyter.Tests/MagicCommandTests.lsmagic.cs]: You can grab the directive's `Command` directly and parse whatever you like:

image

File: src/Microsoft.DotNet.Interactive/Parsing/DirectiveParser.cs (hunk 5 above)
Comment (WhiteBlackGoose) [reply src/Microsoft.DotNet.Interactive/Parsing/DirectiveParser.cs:59]: Aha, I see. Alright, then I'll tighten it up and add tests.

File: src/Microsoft.DotNet.Interactive.Jupyter.Tests/MagicCommandTests.lsmagic.cs (hunk 2 above)
Comment (WhiteBlackGoose) [reply src/Microsoft.DotNet.Interactive.Jupyter.Tests/MagicCommandTests.lsmagic.cs]: Got it. I currently extract it via reflection, I'll switch to this way then.

Comment (WhiteBlackGoose) [reply src/Microsoft.DotNet.Interactive.Jupyter.Tests/MagicCommandTests.lsmagic.cs]: Although it would require having a kernel, which we wanted to avoid for those tests

Comment (jonsequitur) [reply src/Microsoft.DotNet.Interactive.Jupyter.Tests/MagicCommandTests.lsmagic.cs]: If there's a refactoring that makes it easier to test this in isolation, that would be useful. Otherwise, I'd prefer to work through the existing public API rather than use reflection, `InternalsVisibleTo`, etc. even if it pulls in a `Kernel`.

**Comment by WhiteBlackGoose** [comment]
@jonsequitur despite that the parser now works correctly, the reaction to inputs is still strange. Here:
image

In both cases it *applied* the kernel extension, which is good. But in the first case it somehow managed to see those tokens *even though* the parser returned 0 errors (so the parser returns the same thing for both `bla-bla //comment` and `bla-bla // comment` (when there's no space and when there is)).

It seems like there's some other parser for all of it... not sure why you would need two parsers simultaneously...

Another *possible* reason is `KeyValueOptions`. Instead of rewriting it into `DirectiveParseResult`, I just take `ParseResult` and pass there where needed, hoping that `KeyValueOptions` doesn't affect the behaviour. It's hard to rewrite it, because it also abuses some model binder from the CLI API.

To sum up: there are two possible reasons for this behaviour. It's either that there's a second parser which gives a result different from `DirectiveParser`. Or that `KeyValueOptions`, which still take the old `ParseResult` (which contains errors) affect this behaviour. What do you think?

**Comment by jonsequitur** [comment]
Hard to tell. Is the `SubmissionParser` putting the comment into a separate node?

I wouldn't expect `KeyValueOptions` to be involved unless `#!value` is being invoked.

**Comment by WhiteBlackGoose** [comment]
Okay, so we need to do something with this PR. It is clear that this hack... I'd be scared to have it in the actual code. What I made helped to reveal a few problems with the parser, so for example, if you build the code from my PR, you will see, that
```cs
#somemagiccommand --its --arguments // comment with spaces
```
gives an error (whereas the code in `master` would swallow it, at least, the time back when I was working on this PR),
and
```cs
#somemagiccommand --its --arguments //commentwithnospaces
```
would work correctly (unlike the code in `master`, at least that time back). In both cases, as can be seen in the [screenshot](https://github.com/dotnet/interactive/pull/1303#issuecomment-834612948) it calls the kernel extension (which is the goal, *but* there's still a fake parsing error).

I'm not sure I'm able to do something with it without rewriting a lot of things... but maybe this experience will help someone to fix it in a different way.

@jonsequitur if you don't want to have this PR opened, you can close it.
